beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,237 @@
+      node["simulation_dataset"][cycle]["particles"]["current"]["weights"].push_back(
+          static_cast<double>(std::get<1>(current)));
+    }
+
+    for (const auto& propagate : data.propagate) {
+      node["simulation_dataset"][cycle]["particles"]["propagate"]["states"].push_back(std::get<0>(propagate));
+      node["simulation_dataset"][cycle]["particles"]["propagate"]["weights"].push_back(
+          static_cast<double>(std::get<1>(propagate)));
+    }
+
+    for (const auto& reweight : data.reweight) {
+      node["simulation_dataset"][cycle]["particles"]["reweight"]["states"].push_back(std::get<0>(reweight));
+      node["simulation_dataset"][cycle]["particles"]["reweight"]["weights"].push_back(
+          static_cast<double>(std::get<1>(reweight)));
+    }
+
+    for (const auto& resample : data.resample) {
+      node["simulation_dataset"][cycle]["particles"]["resample"]["states"].push_back(std::get<0>(resample));
+      node["simulation_dataset"][cycle]["particles"]["resample"]["weights"].push_back(
+          static_cast<double>(std::get<1>(resample)));
+    }
+
+    node["simulation_dataset"][cycle]["estimation"]["mean"] = std::get<0>(data.estimation);
+    node["simulation_dataset"][cycle]["estimation"]["sd"] = std::get<1>(data.estimation);
+  }
+
+  std::ofstream fout("./src/beluga/beluga_tutorial/bags/" + file_name, std::ios::trunc);
+  fout << node;
+}
+
+int main() {
+  // Load params from yaml file
+  TutorialParams tutorial_params;
+  load_params_from_yaml(tutorial_params);
+
+  // Create the TutorialDataset object for record the simulation data
+  TutorialDataset tutorial_dataset;
+
+  // Generate a random map
+  std::uniform_int_distribution landmark_distribution{0, tutorial_params.map_size};
+  auto landmark_map = beluga::views::sample(landmark_distribution) |
+                      ranges::views::take_exactly(tutorial_params.number_of_doors) | ranges::to<LandmarkMapVector>;
+  landmark_map |= ranges::actions::sort | ranges::actions::unique;
+
+  tutorial_dataset.landmark_map = landmark_map;
+
+  // Generate particles
+  std::normal_distribution<double> initial_distribution(
+      tutorial_params.sim_initial_pose, tutorial_params.initial_pose_sd);
+  auto particles = beluga::views::sample(initial_distribution) |
+                   ranges::views::transform(beluga::make_from_state<Particle>) |
+                   ranges::views::take_exactly(tutorial_params.number_of_particles) | ranges::to<beluga::TupleVector>;
+
+  // Execute the particle filter using beluga
+  int current_pose{tutorial_params.sim_initial_pose};
+  for (auto n = 0; n < tutorial_params.sim_number_of_cycles; n++) {
+    // Check if the simulation is out of bounds
+    if (current_pose > tutorial_params.map_size) {
+      break;
+    }
+
+    // Prepare the data to be saved
+    TutorialSimData tutorial_sim_data;
+
+    // Save ground truth
+    current_pose += (tutorial_params.sim_velocity * tutorial_params.sim_dt);
```

Comment:
The ground truth should have and aleatory movement with some probability (a random variable with some probability to move not move), this is important to validate de localization algorithm

Reply:
some context in https://github.com/AtsushiSakai/PythonRobotics/blob/master/Localization/particle_filter/particle_filter.py#L20-L25

Reply:
Here what you are seeing is the ground truth, look at the line 179 of the code: `int translation_param = generateRandomInt(0, tutorial_params.translation_sigma);` this represent the noise you are talking about.